### ISSUE TITLE
Can not reproduce the same result as you present in OGB-MAG dataset

### ISSUE THREAD

**Issue by eddiegaoo** [facebookresearch/NARS#6]
I used the same setting as you to train OGB-MAG and used the example relation subsets you gave, but the best accuracy I got us only 0.2788. I wonder if there is anything wrong.
```
[INFO] Epoch 840, Times(s): 20.3544, Accuracy: Train 0.3018, Val 0.2757, Test 0.2839
[INFO] Epoch 850, Times(s): 20.1501, Accuracy: Train 0.3021, Val 0.2749, Test 0.2838
[INFO] Epoch 860, Times(s): 20.8749, Accuracy: Train 0.3023, Val 0.2756, Test 0.2851
[INFO] Epoch 870, Times(s): 21.1500, Accuracy: Train 0.3022, Val 0.2754, Test 0.2834
[INFO] Epoch 880, Times(s): 21.2005, Accuracy: Train 0.3023, Val 0.2768, Test 0.2853
[INFO] Epoch 890, Times(s): 21.1369, Accuracy: Train 0.3022, Val 0.2758, Test 0.2842
[INFO] Epoch 900, Times(s): 20.8746, Accuracy: Train 0.3029, Val 0.2763, Test 0.2852
[INFO] Epoch 910, Times(s): 20.9197, Accuracy: Train 0.3028, Val 0.2765, Test 0.2849
[INFO] Epoch 920, Times(s): 20.6873, Accuracy: Train 0.3026, Val 0.2761, Test 0.2839
[INFO] Epoch 930, Times(s): 20.6163, Accuracy: Train 0.3028, Val 0.2771, Test 0.2845
[INFO] Epoch 940, Times(s): 20.7486, Accuracy: Train 0.3033, Val 0.2773, Test 0.2851
[INFO] Epoch 950, Times(s): 20.4859, Accuracy: Train 0.3031, Val 0.2774, Test 0.2853
[INFO] Epoch 960, Times(s): 21.1654, Accuracy: Train 0.3028, Val 0.2788, Test 0.2868
[INFO] Epoch 970, Times(s): 20.6524, Accuracy: Train 0.3030, Val 0.2759, Test 0.2850
[INFO] Epoch 980, Times(s): 20.6713, Accuracy: Train 0.3028, Val 0.2756, Test 0.2843
[INFO] Epoch 990, Times(s): 20.8252, Accuracy: Train 0.3028, Val 0.2783, Test 0.2865
[INFO] Epoch 1000, Times(s): 20.6521, Accuracy: Train 0.3025, Val 0.2760, Test 0.2839
[INFO] Best Epoch 960,Test 0.2788
Namespace(R=5, batch_size=50000, cpu_preprocess=False, data_dir=None, dataset='mag', dropout=0.5, eval_batch_size=50000, eval_every=10, ff_layer=2, gpu=0, input_dropout=False, lr=0.001, num_epochs=100    0, num_hidden=512, seed=None, use_emb='TransE_mag', use_relation_subsets='sample_relation_subsets/examples/mag', weight_decay=0)

```

**Comment by lingfanyu** [comment]
Hi @eddiegaoo ,

Thanks for your interest in this project.

It looks like the model is not training at all. Did you also follow the instructions in [graph_emb](https://github.com/facebookresearch/NARS/tree/main/graph_embed) to generate the TransE embedding?

Also, what's the OGB version you are using? Would you be able to evaluate the dataset with the same python packages that I used:
- torch==1.5.1+cu101
- dgl-cu101==0.4.3.post2
- ogb==1.2.1
- dglke==0.1.0

I also prepared [docker instructions](https://github.com/facebookresearch/NARS/tree/main/docker) for easy reproducing the experiments in the same environment as mine. Can you try evaluating in the docker container? I used AWS G4dn instance which has a T4 GPU.

**Comment by eddiegaoo** [comment]
Thanks for reply. I followed the instructions in graph_emb to generate the TransE embedding, and the OGB I used is 1.2.1 and the only differences in python packages is that I used torch ==1.9.1+cu111. Because if I use torch==1.5.1+cu101 there will always be a bug related to CUDA because the CUDA version of the server I used is CUDA11.2.

**Comment by eddiegaoo** [comment]
```
/home/data/sgaoar/NARS/TransE_mag ⇨  ll
total 3.7G
-rw------- 1 sgaoar 2.2G Sep 30 15:58 author.pt
-rw------- 1 sgaoar 118M Sep 30 15:58 field_of_study.pt
-rw------- 1 sgaoar  18M Sep 30 15:58 institution.pt
-rw------- 1 sgaoar 1.5G Sep 30 15:58 paper.pt
```
I retrained the TransE embedding again and the result is still not promising when just training first 200 epochs as you can see below. 
```
python3 train.py --dataset mag --use-emb TransE_mag --R 5     --use-relation-subset sample_relation_subsets/examples/mag     --eval-batch-size 50000 --num-hidden 512 --lr 0.001 --batch-s 50000     --dropout 0.5 --num-epochs 1000
Using backend: pytorch
Namespace(R=5, batch_size=50000, cpu_preprocess=False, data_dir=None, dataset='mag', dropout=0.5, eval_batch_size=50000, eval_every=10, ff_layer=2, gpu=0, input_dropout=False, lr=0.001, num_epochs=500, num_hidden=512, seed=None, use_emb='TransE_mag', use_relation_subsets='sample_relation_subsets/examples/mag', weight_decay=0)
torch.Size([736389, 128])
torch.Size([128, 256])
Randomly project paper feature from dimension 128 to 256
Reading Relation Subsets:
['writes', 'cites']
['affiliated_with', 'has_topic']
['affiliated_with', 'writes']
['affiliated_with', 'writes', 'cites', 'has_topic']
['writes']
['affiliated_with', 'cites']
['cites', 'has_topic']
['affiliated_with', 'writes', 'cites']
Start generating features for each sub-metagraph:
['writes', 'cites']
['affiliated_with', 'has_topic']
['affiliated_with', 'writes']
['affiliated_with', 'writes', 'cites', 'has_topic']
['writes']
['affiliated_with', 'cites']
['cites', 'has_topic']
['affiliated_with', 'writes', 'cites']
Done preprocessing
[INFO] Namespace(R=5, batch_size=50000, cpu_preprocess=False, data_dir=None, dataset='mag', dropout=0.5, eval_batch_size=50000, eval_every=10, ff_layer=2, gpu=0, input_dropout=False, lr=0.001, num_epochs=500, num_hidden=512, seed=None, use_emb='TransE_mag', use_relation_subsets='sample_relation_subsets/examples/mag', weight_decay=0)
[INFO] new input size: 8 256
[INFO] # Params: 4130149
[INFO] Epoch 10, Times(s): 21.5723, Accuracy: Train 0.2357, Val 0.2315, Test 0.2480
[INFO] Epoch 20, Times(s): 22.3636, Accuracy: Train 0.2504, Val 0.2406, Test 0.2529
[INFO] Epoch 30, Times(s): 20.7159, Accuracy: Train 0.2582, Val 0.2492, Test 0.2593
[INFO] Epoch 40, Times(s): 20.9450, Accuracy: Train 0.2643, Val 0.2524, Test 0.2619
[INFO] Epoch 50, Times(s): 21.0134, Accuracy: Train 0.2685, Val 0.2590, Test 0.2665
[INFO] Epoch 60, Times(s): 20.9920, Accuracy: Train 0.2721, Val 0.2621, Test 0.2710
[INFO] Epoch 70, Times(s): 20.5590, Accuracy: Train 0.2752, Val 0.2632, Test 0.2706
[INFO] Epoch 80, Times(s): 21.0830, Accuracy: Train 0.2779, Val 0.2642, Test 0.2740
[INFO] Epoch 90, Times(s): 21.1757, Accuracy: Train 0.2789, Val 0.2660, Test 0.2736
[INFO] Epoch 100, Times(s): 21.3456, Accuracy: Train 0.2814, Val 0.2690, Test 0.2764
[INFO] Epoch 110, Times(s): 21.3965, Accuracy: Train 0.2825, Val 0.2656, Test 0.2734
[INFO] Epoch 120, Times(s): 21.4828, Accuracy: Train 0.2847, Val 0.2672, Test 0.2741
[INFO] Epoch 130, Times(s): 20.7893, Accuracy: Train 0.2852, Val 0.2663, Test 0.2738
[INFO] Epoch 140, Times(s): 20.8501, Accuracy: Train 0.2864, Val 0.2660, Test 0.2736
[INFO] Epoch 150, Times(s): 20.6377, Accuracy: Train 0.2874, Val 0.2679, Test 0.2753
[INFO] Epoch 160, Times(s): 20.9065, Accuracy: Train 0.2886, Val 0.2681, Test 0.2761
[INFO] Epoch 170, Times(s): 20.7034, Accuracy: Train 0.2893, Val 0.2720, Test 0.2813
[INFO] Epoch 180, Times(s): 21.0184, Accuracy: Train 0.2901, Val 0.2711, Test 0.2785
[INFO] Epoch 190, Times(s): 20.9075, Accuracy: Train 0.2913, Val 0.2734, Test 0.2815
[INFO] Epoch 200, Times(s): 20.5930, Accuracy: Train 0.2915, Val 0.2684, Test 0.2759
```

**Comment by lingfanyu** [comment]
Everything seems fine from the log you shared. I actually have no idea why the model training is not making progress.

Can you also share the log of training TransE embedding with DGLKE? Is the training of embedding converging?

RE: CUDA version issue
If you use the Dockerfile provided in this project, it will pull the offical `nvidia/cuda:10.1-cudnn7-devel-ubuntu18.04` image and creates a container that has cuda-10.1

**Comment by eddiegaoo** [comment]
```
[proc 0][Train](350000/400000) average regularization: 0.0004413209732651012
[proc 0][Train] 10000 steps take 44.080 seconds
[proc 0]sample: 8.615, forward: 14.279, backward: 16.938, update: 4.210
[proc 0][Train](360000/400000) average pos_loss: 0.2650887187063694
[proc 0][Train](360000/400000) average neg_loss: 0.35795491606593133
[proc 0][Train](360000/400000) average loss: 0.3115218173161149
[proc 0][Train](360000/400000) average regularization: 0.0004432984299492091
[proc 0][Train] 10000 steps take 48.539 seconds
[proc 0]sample: 8.385, forward: 14.185, backward: 21.733, update: 4.201
[proc 0][Train](370000/400000) average pos_loss: 0.26343531077057125
[proc 0][Train](370000/400000) average neg_loss: 0.35599396863877775
[proc 0][Train](370000/400000) average loss: 0.30971463962197304
[proc 0][Train](370000/400000) average regularization: 0.00044516465442429764
[proc 0][Train] 10000 steps take 46.897 seconds
[proc 0]sample: 8.276, forward: 14.150, backward: 20.256, update: 4.180
[proc 0][Train](380000/400000) average pos_loss: 0.2616141674682498
[proc 0][Train](380000/400000) average neg_loss: 0.35404962970763443
[proc 0][Train](380000/400000) average loss: 0.3078318986088037
[proc 0][Train](380000/400000) average regularization: 0.00044690257218608166
[proc 0][Train] 10000 steps take 40.813 seconds
[proc 0]sample: 10.410, forward: 14.190, backward: 12.004, update: 4.174
[proc 0][Train](390000/400000) average pos_loss: 0.25437170367836953
[proc 0][Train](390000/400000) average neg_loss: 0.34647522833645344
[proc 0][Train](390000/400000) average loss: 0.3004234659910202
[proc 0][Train](390000/400000) average regularization: 0.0004507235618541017
[proc 0][Train] 10000 steps take 47.759 seconds
[proc 0]sample: 8.534, forward: 14.215, backward: 20.778, update: 4.196
[proc 0][Train](400000/400000) average pos_loss: 0.25327623606026173
[proc 0][Train](400000/400000) average neg_loss: 0.344777174448967
[proc 0][Train](400000/400000) average loss: 0.29902670533806086
[proc 0][Train](400000/400000) average regularization: 0.0004521965909516439
[proc 0][Train] 10000 steps take 49.062 seconds
[proc 0]sample: 8.606, forward: 14.573, backward: 21.584, update: 4.262
proc 0 takes 1670.085 seconds
training takes 1670.0866401195526 seconds
```
Here is the log of training TransE embedding with DGLKE. But I don't know whether this training of embedding is converging or not from the log.

**Comment by lingfanyu** [comment]
I just rented a g4dn.4xlarge instance on AWS, built the docker image from scratch, and ran the MAG example. Here is a screenshot of the first 50 epochs.
![image](https://user-images.githubusercontent.com/11624869/135704571-7c2287a8-ef45-41c3-ab6e-c2c4ae3f88c3.png)

After the first 10 epochs, my training/validation/testing accuracy is much higher than yours.

I strongly suggest that you install docker to use the same environment as mine. It could be possible that DGL 0.4.3 depends on some old pytorch feature, and is not behaving as expected when used with the latest version of pytorch. And then you should also re-train the TransE embedding before running the MAG example.

**Comment by eddiegaoo** [comment]
> I just rented a g4dn.4xlarge instance on AWS, built the docker image from scratch, and ran the MAG example. Here is a screenshot of the first 50 epochs. ![image](https://user-images.githubusercontent.com/11624869/135704571-7c2287a8-ef45-41c3-ab6e-c2c4ae3f88c3.png)
> 
> After the first 10 epochs, my training/validation/testing accuracy is much higher than yours.
> 
> I strongly suggest that you install docker to use the same environment as mine. It could be possible that DGL 0.4.3 depends on some old pytorch feature, and is not behaving as expected when used with the latest version of pytorch. And then you should also re-train the TransE embedding before running the MAG example.

Appreciate it!  I will give it a try and get back to you soon.

**Comment by eddiegaoo** [comment]
```
[INFO] Epoch 10, Times(s): 18.0807, Accuracy: Train 0.4110, Val 0.4028, Test 0.4071
[INFO] Epoch 20, Times(s): 18.1073, Accuracy: Train 0.4641, Val 0.4468, Test 0.4466
[INFO] Epoch 30, Times(s): 18.2643, Accuracy: Train 0.4895, Val 0.4617, Test 0.4595
[INFO] Epoch 40, Times(s): 18.2079, Accuracy: Train 0.5073, Val 0.4744, Test 0.4698
[INFO] Epoch 50, Times(s): 18.2138, Accuracy: Train 0.5231, Val 0.4799, Test 0.4719
[INFO] Epoch 60, Times(s): 18.2428, Accuracy: Train 0.5355, Val 0.4856, Test 0.4781
[INFO] Epoch 70, Times(s): 18.2305, Accuracy: Train 0.5462, Val 0.4940, Test 0.4857
[INFO] Epoch 80, Times(s): 18.1294, Accuracy: Train 0.5557, Val 0.4962, Test 0.4867
[INFO] Epoch 90, Times(s): 18.1927, Accuracy: Train 0.5659, Val 0.5003, Test 0.4900
[INFO] Epoch 100, Times(s): 18.1622, Accuracy: Train 0.5723, Val 0.5055, Test 0.4931
[INFO] Epoch 110, Times(s): 18.1693, Accuracy: Train 0.5818, Val 0.5040, Test 0.4903
[INFO] Epoch 120, Times(s): 18.1901, Accuracy: Train 0.5895, Val 0.5044, Test 0.4931
[INFO] Epoch 130, Times(s): 18.2559, Accuracy: Train 0.5950, Val 0.5096, Test 0.4958
[INFO] Epoch 140, Times(s): 18.1847, Accuracy: Train 0.6034, Val 0.5114, Test 0.4974
[INFO] Epoch 150, Times(s): 18.1430, Accuracy: Train 0.6103, Val 0.5155, Test 0.4996
[INFO] Epoch 160, Times(s): 18.3362, Accuracy: Train 0.6163, Val 0.5144, Test 0.5012
[INFO] Epoch 170, Times(s): 18.1042, Accuracy: Train 0.6215, Val 0.5164, Test 0.5039
[INFO] Epoch 180, Times(s): 18.1757, Accuracy: Train 0.6278, Val 0.5173, Test 0.5046
[INFO] Epoch 190, Times(s): 18.2002, Accuracy: Train 0.6311, Val 0.5211, Test 0.5085
[INFO] Epoch 200, Times(s): 18.1885, Accuracy: Train 0.6378, Val 0.5176, Test 0.5052
[INFO] Epoch 210, Times(s): 18.2973, Accuracy: Train 0.6416, Val 0.5177, Test 0.5042
[INFO] Epoch 220, Times(s): 18.1865, Accuracy: Train 0.6451, Val 0.5211, Test 0.5071

```
Finally I make it work. Thank you for keeping offering the help! It really means a lot. Now that the problem is solved, I will close this issue. Wish you all the best in the future research!

**Comment by lingfanyu** [comment]
@eddiegaoo It's great to know that the code is working. Thanks!
